### PR TITLE
Add warning for TypeScript set up issues and fix Babel config

### PR DESCRIPTION
**Pull request by mrmckeb**
Issue: #10943

## What I did
- Added a warning for projects that have TypeScript files, but not TypeScript as a dependency.
- Addressed an issue where optional chaining and nullish coalescing may not be processed correctly.

## How to test
Tests have been added for the CLI.

### CODE REVIEW

File: lib/cli/src/warn.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import globby from 'globby';
+import { logger } from '@storybook/node-logger';
+
+interface Options {
+  hasTSDependency: boolean;
+}
+
+export const warn = ({ hasTSDependency }: Options) => {
+  if (!hasTSDependency) {
+    const hasTSFiles = !!globby.sync(['**/*.(ts|tsx)', '!**/node_modules', '!**/*.d.ts']).length;
```
Comment (shilman) [inline lib/cli/src/warn.ts:10]: `'**/*.(ts|tsx)'` is not a proper glob 😄   See https://github.com/storybookjs/storybook/pull/10926

Comment (shilman) [reply lib/cli/src/warn.ts:10]: This is what I used in codemod if it's useful:

https://github.com/storybookjs/storybook/blob/next/lib/codemod/src/index.js#L45

**Review by shilman — CHANGES_REQUESTED** [review]
One comment otherwise LGTM

File: lib/cli/src/warn.ts (hunk 1 above)
Comment (mrmckeb) [reply lib/cli/src/warn.ts:10]: According to the docs here:
https://github.com/isaacs/minimatch#usage
https://github.com/sindresorhus/globby#patterns

This is supported, and is called `extglob`. Maybe you can give me some more info around the change? I didn't see any docs linked in the PR referenced.

Comment (ndelangen) [reply lib/cli/src/warn.ts:10]: `const hasTSFiles = !!globby.sync(['**/*.(ts|tsx)', '!**/node_modules', '!**/*.d.ts']).length;`

should be

`const hasTSFiles = !!globby.sync(['**/*.@(ts|tsx)', '!**/node_modules', '!**/*.d.ts']).length;`

**Review by ndelangen — CHANGES_REQUESTED** [review]

**Comment by mrmckeb** [comment]
OK, I found the reference here:
https://www.npmjs.com/package/extglob#extglob-cheatsheet

File: lib/cli/src/warn.test.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import globby from 'globby';
```
Comment (ndelangen) [inline lib/cli/src/warn.test.ts:1]: any reason to add a dep when we already depend on `glob`?

**Review by ndelangen — APPROVED** [review]

File: lib/cli/src/warn.test.ts (hunk 2 above)
Comment (mrmckeb) [reply lib/cli/src/warn.test.ts:1]: In CLI, there was no dependency. Globby seemed more reliable from my (quick) testing of the two, and supports arrays. This implementation was _lifted_ from the same check in CRA.

**Pull request merged**